File: research/framework-bench/apps/blazor/Services/SettingsService.cs
namespace BillSplitter.Services;

// Shared/global state as a DI singleton with a change notification — the
// idiomatic Blazor cross-component shared-state pattern. Components subscribe to
// OnChange and call StateHasChanged.
public class SettingsService
{
    public string Theme { get; private set; } = "light";
    public bool RoundUp { get; private set; }

    public event Action? OnChange;

    public void ToggleTheme()
    {
        Theme = Theme == "dark" ? "light" : "dark";
        OnChange?.Invoke();
    }

    public void ToggleRoundUp()
    {
        RoundUp = !RoundUp;
        OnChange?.Invoke();
    }
}

File: research/framework-bench/apps/angular/src/app/settings.service.ts
import { Injectable, signal } from "@angular/core";

// Shared/global state as an injectable singleton service holding signals —
// Angular's idiomatic cross-component shared-state mechanism.
@Injectable({ providedIn: "root" })
export class SettingsService {
  readonly theme = signal<"light" | "dark">("light");
  readonly roundUp = signal(false);

  toggleTheme() {
    this.theme.update((t) => (t === "dark" ? "light" : "dark"));
  }
  toggleRoundUp() {
    this.roundUp.update((r) => !r);
  }
}
